### PR TITLE
Harmonize SLD height

### PR DESCRIPTION
**Pull request by Meklo**

### CODE REVIEW

**Comment by EstherDarkish** [comment]
OK, code is fine by me.

**Comment by Meklo** [comment]
> OK, code is fine by me.

I reworked the code in order to implement a more robust solution

**Comment by Meklo** [comment]
Note : I tried to incorporate the initialHeight value into the displayedSLD array but the value was too often overriden due to the way displayedSLD is updated. For reliability reasons I created a new array, displayedSldHeights, storing intialHeight values.

**Review by EstherDarkish — APPROVED** [review]

File: src/components/diagrams/singleLineDiagram/single-line-diagram.js [hunk 1]
```diff
@@ -756,7 +780,7 @@ const SingleLineDiagram = forwardRef((props, ref) => {
                     <AlertInvalidNode noMargin={true} />
                 </Box>
             ) : (
-                <Box position="relative">
```
Comment (etiennehomer) [inline src/components/diagrams/singleLineDiagram/single-line-diagram.js:759]: Why removing ? It's useless ?

Comment (Meklo) [reply src/components/diagrams/singleLineDiagram/single-line-diagram.js:759]: That change is what enable the full screen buttons to be aligned. Previously its position was relative to the svg container, by removing this line it is now relative to sld pane

Comment (etiennehomer) [reply src/components/diagrams/singleLineDiagram/single-line-diagram.js:759]: Ah yes, I see

File: src/components/diagrams/singleLineDiagram/single-line-diagram.js [hunk 2]
```diff
@@ -326,6 +328,24 @@ const SingleLineDiagram = forwardRef((props, ref) => {
     const [svgFinalWidth, setSvgFinalWidth] = useState();
     const [svgFinalHeight, setSvgFinalHeight] = useState();
 
+    useEffect(() => {
+        if (finalPaperHeight) {
+            setDisplayedSldHeights((displayedSldHeights) => {
+                if (
+                    displayedSldHeights.filter((sld) => sld.id === sldId)
+                        .length > 0
```
Comment (etiennehomer) [inline src/components/diagrams/singleLineDiagram/single-line-diagram.js:336]: displayedSldHeights.find(sld => sld.id === sldId) ?

**Comment by etiennehomer** [comment]
There's a problem when we close the last SLD and when the "new last one" is small in height

File: src/components/diagrams/singleLineDiagram/single-line-diagram.js (hunk 2 above)
Comment (Meklo) [reply src/components/diagrams/singleLineDiagram/single-line-diagram.js]: I removed this bit

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 3]
```diff
@@ -318,6 +321,37 @@ export function SingleLineDiagramPane({
 
     const displayedIds = new Set(displayedSLD.map(({ id }) => id));
     const minimized = views.filter(({ id }) => !displayedIds.has(id));
+    const [computedHeight, setComputedHeight] = useState();
+
+    useEffect(() => {
+        let displayedSldHeights_ = displayedSldHeightsRef.current;
+        displayedSldHeights_ = displayedSldHeights_?.filter((displayedHeight) =>
+            viewsRef.current
```
Comment (etiennehomer) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:329]: use views here as you have it in the dependencies

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 4]
```diff
@@ -318,6 +321,37 @@ export function SingleLineDiagramPane({
...
+                .map((sld) => sld.id)
+                .includes(displayedHeight.id)
+        );
+
+        setDisplayedSldHeights(displayedSldHeights_);
+    }, [views]);
+
+    useEffect(() => {
+        if (displayedSldHeights.length === 0) {
+            setComputedHeight();
```
Comment (etiennehomer) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:340]: We need this code ?

**Review by etiennehomer — CHANGES_REQUESTED** [review]

**Comment by etiennehomer** [comment]
For a SVG that is smaller than the other opened ones, the bottom of the pane is not usable for the svg navigation. 
Not only the Paper height should be to computedHeight, the SVG height also

File: src/components/diagrams/singleLineDiagram/single-line-diagram.js [hunk 5]
```diff
@@ -326,6 +328,17 @@ const SingleLineDiagram = forwardRef((props, ref) => {
...
+                return [
+                    ...displayedSldHeights.filter((sld) => sld.id !== sldId),
+                    { id: sldId, initialHeight: finalPaperHeight },
+                ];
+            });
+        }
+    }, [finalPaperHeight, setDisplayedSldHeights, sldId]);
+
     useLayoutEffect(() => {
         const sizes = computePaperAndSvgSizesIfReady(
```
Comment (etiennehomer) [inline src/components/diagrams/singleLineDiagram/single-line-diagram.js:343]: Something like setSvgFinalHeight(computedHeight); could work for all the pane to be navigable with the SVG

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js (hunk 3 above)
Comment (Meklo) [reply src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js]: Done

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js (hunk 4 above)
Comment (Meklo) [reply src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js]: Not really, I removed it

File: src/components/diagrams/singleLineDiagram/single-line-diagram.js (hunk 5 above)
Comment (Meklo) [reply src/components/diagrams/singleLineDiagram/single-line-diagram.js:343]: It was causing visual glitches if I implemented it through setSvgFinalHeight, I found a way by directly setting the svg height attribute line 589

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 6]
```diff
@@ -318,6 +321,33 @@ export function SingleLineDiagramPane({
...
+
+    useEffect(() => {
+        const initialHeights = [
+            ...displayedSldHeights.map(
+                (displayedHeight) => displayedHeight.initialHeight
+            ),
+        ];
+        if (initialHeights.length > 0) {
+            const newComputedHeight = Math.max(...initialHeights);
+            if (newComputedHeight && !isNaN(newComputedHeight)) {
```
Comment (etiennehomer) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:346]: I think you can remove !isNaN(newComputedHeight)

**Review by etiennehomer — CHANGES_REQUESTED** [review]

**Comment by etiennehomer** [comment]
Tests OK

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 7]
```diff
@@ -318,6 +321,33 @@ export function SingleLineDiagramPane({
 
     const displayedIds = new Set(displayedSLD.map(({ id }) => id));
     const minimized = views.filter(({ id }) => !displayedIds.has(id));
+    const [computedHeight, setComputedHeight] = useState();
+
+    useEffect(() => {
+        let displayedSldHeights_ = displayedSldHeightsRef.current;
+        displayedSldHeights_ = displayedSldHeights_?.filter((displayedHeight) =>
```
[the remark this answers is not in the document]
Comment (Meklo) [reply src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js]: Done

**Review by etiennehomer — APPROVED** [review]

**Pull request merged**